src/cdomains/apron/linearTwoVarEqualityDomain.apron.ml

```diff
@@ -0,0 +1,727 @@
+          begin match b with
+            | Add -> List.concat [convert_texpr e1; convert_texpr e2]
+            | Sub -> List.concat [convert_texpr e1; negate (convert_texpr e2)]
+            | Mul -> multiply (convert_texpr e1) (convert_texpr e2)
+            | _ -> raise NotLinearExpr end
+      end
+    in match convert_texpr texp with
+    | exception NotLinearExpr -> None
+    | x -> Some(x)
+
+  let get_coeff (t: t) texp =
+    (*Parses a Texpr to obtain a (variable, offset) pair to repr. a sum of a variable and an offset.
+      Returns None if the expression is not a sum between a variable (without coefficient) and a constant.
+    *)
+    let exception Not2VarExpr in
+    let sum_next_coefficient (var, current_var_offset, curr_offset) (next_coeff, next_var) =
+      begin match next_var with
+        | None -> (* this element represents a constant offset *)
+          (var, current_var_offset, Z.(curr_offset + next_coeff))
+        | Some _ -> (* this element represents a variable with a coefficient
+                              -> it must be always the same variable, else it's not a two-variable equality*)
+          begin if Option.is_none var || next_var = var then
+              (next_var, Z.(current_var_offset + next_coeff), curr_offset)
+            else raise Not2VarExpr end
+      end in
+    let sum_coefficients summands_list_opt =
+      Option.map (List.fold_left sum_next_coefficient (None, Z.zero, Z.zero)) summands_list_opt
+    in
+    match sum_coefficients (get_coeff_vec t texp) with
+    | exception _ -> None
+    | None -> None
+    | Some (var, var_coeff, offset) ->
+      if Option.is_none var then Some (None, offset)
+      else if Z.(var_coeff = one) then Some (var, offset)
+      else None
+
+
+  let get_coeff t texp = timing_wrap "coeff_vec" (get_coeff t) texp
+
+  let abstract_exists var t = match t.d with
+    | Some d -> {t with d = Some (EArray.reduce_col d (Environment.dim_of_var t.env var))}
+    | None -> t (* there are no  variables in the current environment *)
+
+  let assign_const t var const = match t.d with
+    | None -> t
+    | Some t_d -> let d = EArray.copy t_d in d.(var) <- (None, const);  {d = Some d; env = t.env}
+
+  let subtract_const_from_var t var const =
+    match t.d with
+    | None -> t
+    | Some t_d -> let d = EArray.copy t_d in
+      let subtract_const_from_var_for_single_equality index element =
+        let (eq_var_opt, off2) = d.(index) in
+        if index = var then
+          match eq_var_opt with
+          | None -> d.(index) <- (None, Z.(off2 + const))
+          | Some eq_var -> begin if eq_var <> index then d.(index) <- (eq_var_opt, Z.(off2 + const)) end
+        else
+          begin match eq_var_opt with
+            | Some eq_var ->
+              if eq_var = var then d.(index) <- (eq_var_opt, Z.(off2 - const))
+            | None -> ()
+          end
+      in
+      EArray.iteri (subtract_const_from_var_for_single_equality) d; {d = Some d; env = t.env}
```

Comment:
Is there a particular reason for iterating over all array elements in all cases? 
iterating makes sense, if var can be a reference variable, because then it can appear on the right-hand side of equalities. However, this is only the case if (Some var,0)=d.(var), or did I oversee something?

Reply:
That's true, I hadn't thought of that.